File: javascript/ankit_kanyal/3Sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function (nums) {
  if (nums.length < 3) return [];

  let ans = [];
  nums.sort((a, b) => a - b);
  for (let i = 0; i < nums.length - 1; i++) {
    if (i > 0 && nums[i] === nums[i - 1]) continue;

    let j = i + 1;
    let k = nums.length - 1;

    while (j < k) {
      console.log(i, j, k);
      let sum = nums[i] + nums[j] + nums[k];
      console.log("Sum " + sum);
      if (sum === 0) {
        ans.push([nums[i], nums[j], nums[k]]);
        let val1 = nums[j];
        while (j < k && nums[j] == val1) j++;

        let val2 = nums[k];
        while (j < k && nums[k] == val2) k--;
      } else if (sum < 0) j++;
      else if (sum > 0) k--;
    }
  }

  return ans;
};

let arr = [1, -1, -1, 0];

console.log(threeSum(arr));

File: Python/Shubham Sigdar/3Sum.py
# 1. Brute force Error-Time Limit Exceeded
# Time complexity-o(n^3)
class Solution(object):
    def threeSum(self, nums):
        
        n=len(nums)
        target=0
        arr=[]
        result=[]
        for i,el1 in enumerate(nums):
            for j,el2 in enumerate(nums):
                for k,el3 in enumerate(nums):
                    if el1+el2+el3 == target and i!=j and i!=k and j!=k:
                        #sort 
                        arr=sorted([el1,el2,el3])
                       #remove duplicate elements
                        if arr not in result:
                            result.append( sorted( [el1 ,el2 , el3] ))
        

        return result


# 2. 2 pointer method

def solve(n, nums):
        res = []
        nums.sort()
        for i, a in enumerate(nums):
            #we don't wanna use same value in a same position
            if i > 0 and a == nums[i - 1]: #a == nums[i - 1] same element before
                continue
            
            l, r = i + 1, len(nums) - 1
            while l < r:
                threeSum = a + nums[l] + nums[r]
                if threeSum > 0:
                    r -= 1
                elif threeSum < 0:
                    l += 1
                else:
                    res.append([a, nums[l], nums[r]])
                    #update your pointer
                    while l < r and nums[l] == nums[l+1]: 
                        l += 1 
                    while l < r and nums[r] == nums[r-1]:
                        r -= 1 
                    l += 1; r -= 1

        return res